lib/PuppeteerSharp.Tests/Page/GotoTests.cs

```diff
@@ -0,0 +1,228 @@
+            Assert.Equal(HttpStatusCode.OK, response.Status);
+        }
+
+        [Fact]
+        public async Task ShouldWorkWhenNavigatingTo404()
+        {
+            var response = await Page.GoToAsync(TestConstants.ServerUrl + "/not-found");
+            Assert.Equal(HttpStatusCode.NotFound, response.Status);
+        }
+
+        [Fact]
+        public async Task ShouldReturnLastResponseInRedirectChain()
+        {
+            Server.SetRedirect("/redirect/1.html", "/redirect/2.html");
+            Server.SetRedirect("/redirect/2.html", "/redirect/3.html");
+            Server.SetRedirect("/redirect/3.html", TestConstants.EmptyPage);
+
+            var response = await Page.GoToAsync(TestConstants.ServerUrl + "/redirect/1.html");
+            Assert.Equal(HttpStatusCode.OK, response.Status);
+            Assert.Equal(TestConstants.EmptyPage, response.Url);
+        }
+
+        [Fact]
+        public async Task ShouldWaitForNetworkIdleToSucceedNavigation()
+        {
+            var responses = new List<TaskCompletionSource<Func<HttpResponse, Task>>>();
+            foreach (var url in new[] { "/fetch-request-a.js", "/fetch-request-b.js", "/fetch-request-c.js", "/fetch-request-d.js" })
+            {
+                Server.SetRoute(url, async context =>
+                {
+                    var taskCompletion = new TaskCompletionSource<Func<HttpResponse, Task>>();
+                    responses.Add(taskCompletion);
+                    var actionResponse = await taskCompletion.Task;
+                    await actionResponse(context.Response);
+                });
+            }
+
+            var initialFetchResourcesRequested = Task.WhenAll(
+                Server.WaitForRequest("/fetch-request-a.js"),
+                Server.WaitForRequest("/fetch-request-b.js"),
+                Server.WaitForRequest("/fetch-request-c.js")
+            );
+            var secondFetchResourceRequested = Server.WaitForRequest("/fetch-request-d.js");
+
+            var navigationFinished = false;
+            var navigationTask = Page.GoToAsync(TestConstants.ServerUrl + "/networkidle.html",
+                new NavigationOptions { WaitUntil = new[] { WaitUntilNavigation.Networkidle0 } })
+                .ContinueWith(res =>
+                {
+                    navigationFinished = true;
+                    return res.Result;
+                });
+
+            var pageLoaded = new TaskCompletionSource<bool>();
+            void WaitPageLoad(object sender, EventArgs e)
+            {
+                pageLoaded.SetResult(true);
+                Page.Load -= WaitPageLoad;
+            }
+            Page.Load += WaitPageLoad;
+            await pageLoaded.Task;
+
+            Assert.False(navigationFinished);
+
+            await initialFetchResourcesRequested;
+
+            Assert.False(navigationFinished);
+
+            while (responses.Count != 3) await Task.Delay(10); // todo: figure out why
```

Comment:
The test was reaching the foreach on line 174 with an empty list, this worked, not sure why (debugging passes since it's slower)
same on line 189